core/tracestack.lua

```diff
@@ -0,0 +1,214 @@
+    local top = traceStack[#traceStack]
+    if not top then
+      -- Stack is empty, there is not much we can do
+      return afterFrame and "after " .. afterFrame:toLocationString() or nil
+    end
+    local info = top:toLocationString()
+    -- Not all stack traces have to carry location information.
+    -- If the top stack trace does not carry it, find a frame which does.
+    -- Then append it, because its information may be useful.
+    if not top.line then
+      for i = #traceStack - 1, 1, -1 do
+        if traceStack[i].line then
+          -- Found a frame which does carry some relevant information.
+          info = info .. " near " .. traceStack[i]:toLocationString(--[[skipFile=]] traceStack[i].file == top.file)
+          break
+        end
+      end
+    end
+    -- Print after, if it is in a relevant file
+    if afterFrame and (not base or afterFrame.file == base.file) then
+      info = info .. " after " .. afterFrame:toLocationString(--[[skipFile=]] true)
+    end
+    return info
+  end,
+
+  -- Push a command frame on to the stack to record the execution trace for debugging.
+  -- Carries information about the command call, not the command itself.
+  -- Must be popped with `pop(returnOfPush)`.
+  pushCommand = function(self, tag, line, column, options, file)
+    if not tag then
+      SU.warn("Tag should be specified for SILE.traceStack:pushCommand", true)
+    end
+    return self:_pushFrame({
+        tag = tag or "???",
+        file = file or SILE.currentlyProcessingFile,
+        line = line,
+        column = column,
+        options = options or {}
+      })
+  end,
+
+
+  -- Push a command frame on to the stack to record the execution trace for debugging.
+  -- Command arguments are inferred from AST content, any item may be overridden.
+  -- Must be popped with `pop(returnOfPush)`.
+  pushContent = function(self, content, tag, line, column, options, file)
+    if type(content) ~= "table" then
+      SU.warn("Content parameter of SILE.traceStack:pushContent must be a table", true)
+      content = {}
+    end
+    tag = tag or content.tag
+    if not tag then
+      SU.warn("Tag should be specified or inferable for SILE.traceStack:pushContent", true)
+    end
+    return self:_pushFrame({
+        tag = tag or "???",
+        file = file or content.file or SILE.currentlyProcessingFile,
+        line = line or content.line,
+        column = column or content.col,
+        options = options or content.attr or {}
+      })
+  end,
+
+  -- Push a text that is going to get typeset on to the stack to record the execution trace for debugging.
+  -- Must be popped with `pop(returnOfPush)`.
+  pushText = function(self, text, line, column, file)
+    return self:_pushFrame({
+        text = text,
+        file = file,
+        line = line,
+        column = column
+      })
+  end,
+
+  -- Internal: Push complete frame
+  _pushFrame = function (self, frame)
+    frame.toLocationString = self._frameToLocationString
```

Comment:
This seems like a bug. Given that there is only ever one instance of `SILE.traceStack`, the function assignment here is always going to return the same thing. As such there is so point in saving a pointer to it with each "frame" on the stack, and calling it can be a bit simpler too.

Am I missing something or can I just factor this out and use `self:toLocationString()` instead of `popped:toLocationString()`.

Reply:
No, this is correct. Think of it as "inheritance lite". All stack frame objects (erm, "tables") contain a function which can print them. In a more OOP style, I'd somewhere create a "base class" with this method and let further frames be instances of it. But that seems like overkill.

Should this line be removed, all instances of `<frame>:toLocationString(<anything>)` would have to be rewritten to `SILE.traceStack._frameToLocationString(<frame>, <anything>)`. As you can see, that is very verbose, so I did it this way.

Additional benefit of this for the future would be that different frames could have different printing methods. Right now, there are two types of frames: command (created by `pushCommand` and `pushContent`) and text (`pushText`). Both have slightly different fields, so they are printed differently, but the difference is tiny compared to the amount of shared code, so the `_frameToLocationString` deals with that internally. However, if in the future we add something like `pushGeneric`, then it might be useful to allow users to provide their own `toLocationString` function.

Your suggested change would currently break this, as:
1. You are calling it without arguments. It expects at least the `self` argument (a frame), to know what it should print.
2. Other code expects `frame.toLocationString` to be a function, not a pre-formatted string (which it returns).

I have thought about storing the formatted frame information in the frame, but it would be slow and not flexible.

Reply:
My sample code was just to let you know where I was talking about, it wasn't meant to be a complete replacement. I see what's going no now, thanks for the explanation.